### PR TITLE
Fix #1119: Adds Bookmarklet support with validation.

### LINKED ISSUE #1119: Add bookmarklet support
### Description: 

Like mentioned in the previous repo, there is a need for the bookmarklet support https://github.com/brave/browser-ios/issues/1724. Thus feature is supported by both safari and firefox on ios.

### PR DESCRIPTION
**Pull request by Brandon-T**
- Adds Bookmarklet support with validation.
- Fixes https://github.com/brave/brave-ios/issues/1119
- Allow JS to run in the URL bar (can be disabled and Bookmarklets will still work - or make it into a preference).
- Security Review: https://github.com/brave/internal/issues/637

When adding a bookmarklet, if you enter into the URL field:

`"javascript:"` and some JS code right after it, it will save as a `Bookmarklet`.
The next time you visit any page, you can execute the bookmarklet and it will modify the web-view/page/content accordingly.

IE (Bookmark):

title: `"view-source"`
url: 
```javascript
javascript:(function(){var a=window.open('about:blank').document;a.write('<!DOCTYPE html><html><head><title>Source of '+location.href+'</title><meta name="viewport" content="width=device-width" /></head><body></body></html>');a.close();var b=a.body.appendChild(a.createElement('pre'));b.style.overflow='auto';b.style.whiteSpace='pre-wrap';b.appendChild(a.createTextNode(document.documentElement.innerHTML))})();
```

Save it as a bookmark. Go to lets say `youtube` for example and tap on bookmarks -> view-source.. It will now execute the above code and you'll be able to view the source-code of youtube for example..

Another one is that you can run JS in the URL bar by typing in the URL bar like:
```javascript
javascript:location.href="https://stackoverflow.com";
```

Note: Brave does not support `view-source` URI natively.. but it does through bookmarklet support if the PR goes through :)

Note: Not sure if it should be integrated like this or in a separate menu or something. No idea, but it's here.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

**TEST PLAN:**
**Test Plan:**

1. Add a new bookmark.
2. Change the url to `javascript:function(){}`
3. It should fail to save.

-

1. Add a new bookmark.
2. Change the url to `javascript:`
3. Should fail to save.

-

1. Add a new bookmark.
2. Change the url to `https://brave.com`
3. It MUST save successfully!
4. Open the bookmark and it should go to the correct url.
5. Any current bookmarks MUST work as well.

-

1. Add a new bookmark.
2. Change the url to `javascript:(function(){ alert("hello"); })()`
3. Save it successfully.
4. Go to any website and open your bookmark.
5. Should show a popup saying `“Hello”`.

ONLY if all the above steps for EVERY case works, this ticket can be marked resolved. If one has failed, we **MUST** file a new bug or message me, as bookmarks are a sensitive thing for users and we don't want to lose any functionality for them.

### CODE REVIEW

**Review by iccub — COMMENTED** [review]
What icon/favicon is showing for a saved bookmarklet?
Can they be added as Favorites too?

**Comment by Brandon-T** [comment]
@iccub It shows this "planet" icon for Bookmarklets.
Yes they can be added as favourites as well since favourites are just bookmarks but on a different screen/folder (IE: Home-screen).

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -870,19 +870,35 @@ class BrowserViewController: UIViewController {
     }
 
     func finishEditingAndSubmit(_ url: URL, visitType: VisitType) {
-        topToolbar.currentURL = url
-        topToolbar.leaveOverlayMode()
+        let javascriptPrefix = "javascript:"
+        if url.absoluteString.hasPrefix(javascriptPrefix) {
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift]: I feel like we should add `isBookmarklet` helper to `URL`, add it to `URLExtensions.swift` and replace all these hasPrefix occurences

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift [hunk 2]
```diff
@@ -235,9 +235,21 @@ class AddEditBookmarkTableViewController: UITableViewController {
         // Saving and udpating is slightly different for each mode.
         switch mode {
         case .addBookmark(_, _):
-            guard let urlString = bookmarkDetailsView.urlTextField?.text,
-                let url = URL(string: urlString) else {
+            guard let urlString = bookmarkDetailsView.urlTextField?.text else {
+                return earlyReturn()
+            }
+            
+            var url: URL! = URL(string: urlString)
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift]: unwrap it, add it to the guard statement few lines above, should work with `var`

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift [hunk 3]
```diff
@@ -235,9 +235,21 @@ class AddEditBookmarkTableViewController: UITableViewController {
...
+            if urlString.hasPrefix("javascript:") {
+                guard let escapedURLString = urlString.addingPercentEncoding(withAllowedCharacters: .URLAllowed), let escapedURL = URL(string: escapedURLString) else {
                     return earlyReturn()
+                }
+                
+                url = escapedURL
+            }
+            
+            guard url != nil else {
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift]: replace with `if url == nil { return }`

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift [hunk 4]
```diff
@@ -258,16 +270,28 @@ class AddEditBookmarkTableViewController: UITableViewController {
                 Bookmark.addFolder(title: title, parentFolder: folder)
             }
         case .editBookmark(let bookmark):
-            guard let urlString = bookmarkDetailsView.urlTextField?.text,
-                let url = URL(string: urlString) else {
+            guard let urlString = bookmarkDetailsView.urlTextField?.text else {
+                return earlyReturn()
+            }
+            
+            var url: URL! = URL(string: urlString)
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift]: unwrap it in guard statement

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift [hunk 5]
```diff
@@ -258,16 +270,28 @@ class AddEditBookmarkTableViewController: UITableViewController {
...
+            if urlString.hasPrefix("javascript:") {
+                guard let escapedURLString = urlString.addingPercentEncoding(withAllowedCharacters: .URLAllowed), let escapedURL = URL(string: escapedURLString) else {
                     return earlyReturn()
+                }
+                
+                url = escapedURL
+            }
+            
+            guard url != nil else {
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift]: replace with `if`

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift [hunk 6]
```diff
@@ -20,10 +21,12 @@ class BookmarkDetailsView: AddEditHeaderView, BookmarkFormFieldsProtocol {
     
     let urlTextField: UITextField? = UITextField().then {
         $0.placeholder = Strings.BookmarkUrlPlaceholderText
-        $0.keyboardType = .URL
+        //$0.keyboardType = .URL
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: I don't think we should remove it, most people is going to add regular bookmarks, not bookmarklets.
Is this keyboard type causing problems with adding bookmarklets?

**Review by iccub — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 7]
```diff
@@ -870,19 +870,35 @@ class BrowserViewController: UIViewController {
     }
 
     func finishEditingAndSubmit(_ url: URL, visitType: VisitType) {
-        topToolbar.currentURL = url
-        topToolbar.leaveOverlayMode()
+        let javascriptPrefix = "javascript:"
```
Comment (diracdeltas) [inline Client/Frontend/Browser/BrowserViewController.swift]: is the URL normalized at this point? for instance `JAVASCRIPT://...` wouldn't match this but would be resolvable as javascript

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift [hunk 8]
```diff
@@ -235,9 +235,21 @@ class AddEditBookmarkTableViewController: UITableViewController {
...
         switch mode {
         case .addBookmark(_, _):
-            guard let urlString = bookmarkDetailsView.urlTextField?.text,
-                let url = URL(string: urlString) else {
+            guard let urlString = bookmarkDetailsView.urlTextField?.text else {
+                return earlyReturn()
+            }
+            
+            var url: URL! = URL(string: urlString)
+            if urlString.hasPrefix("javascript:") {
```
Comment (diracdeltas) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift]: same as https://github.com/brave/brave-ios/pull/1385/files#r313524191

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift [hunk 9]
```diff
@@ -88,6 +94,45 @@ class BookmarkDetailsView: AddEditHeaderView, BookmarkFormFieldsProtocol {
...
-        delegate?.correctValues(validationPassed: validateFields())
+        delegate?.correctValues(validationPassed: validateFields() || validateCodeFields())
+    }
+    
+    private func validateTitle(_ title: String?) -> Bool {
+        guard let title = title else { return false }
+        return !title.isEmpty
+    }
+    
+    private func validateCodeFields() -> Bool {
```
Comment (diracdeltas) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift:108]: this could use a unit test

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift (hunk 6 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: > I don't think we should remove it, most people is going to add regular bookmarks, not bookmarklets.
> Is this keyboard type causing problems with adding bookmarklets?

Yeah. It doesn't let you type in a bookmarklet manually because the URL type doesn't support `"`, `{`, `}`, etc.. :(

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift [hunk 10]
```diff
@@ -88,6 +94,45 @@ class BookmarkDetailsView: AddEditHeaderView, BookmarkFormFieldsProtocol {
...
+        
+        javascriptCode = String(javascriptCode.dropFirst(uriScheme.count))
+        
+        if javascriptCode.isEmpty {
+            return false
+        }
+        
+        // A bookmarklet is considered valid if it's code is valid JS.
+        // Bookmarklets MIGHT invoke some security flaws allowing the user to run arbitrary
+        // JS in the browser though.
```
Comment (diracdeltas) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: maybe specify in the comment that this is running in the webpage context (as opposed to browser context)

assuming this is implemented correctly, the security implications of bookmarklets should only be that they introduce a new way for self-XSS

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 7 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController.swift]: Fixed.

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift (hunk 8 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift]: Fixed.

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift (hunk 9 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift:108]: Fixed. Added a unit test for Bookmarklets.

File: Shared/Extensions/URLExtensions.swift [hunk 11]
```diff
@@ -468,6 +468,39 @@ extension URL {
...
+    
+    public var bookmarkletCodeComponent: String? {
+        return self.absoluteString.bookmarkletCodeComponent
+    }
+}
+
+extension String {
+    public var isBookmarklet: Bool {
+        let url = self.lowercased()
+        return url.hasPrefix("javascript:") && !url.hasPrefix("javascript:/")
```
Comment (diracdeltas) [inline Shared/Extensions/URLExtensions.swift]: FWIW there are potentially other ways to bypass this (embedded tabs, etc): https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet

if firefox has some kind of internal functionality for detecting bookmarklets / js urls, that would be preferable

however i don't think there is any security impact to a URL not being detected as a bookmarklet? IIUC that just means nothing will happen when it is entered.

Comment (Brandon-T) [reply Shared/Extensions/URLExtensions.swift]: @diracdeltas I tried quite a few of these. They don't work.

I create a script that injects a tag into each page and attempts some of the above:

```javascript
var img = document.createElement("img");
img.src = "jav  ascript:window.location.href='https://brave.com';";
document.body.appendChild(img);
```

```javascript
var img = document.createElement("img");
img.src = "jav\tascript:window.location.href='https://brave.com';";
document.body.appendChild(img);
```

```javascript
var img = document.createElement("img");
img.src = "jav\nascript:window.location.href='https://brave.com';";
document.body.appendChild(img);
```

```javascript
var img = document.createElement("img");
img.src = "jav&#x09;ascript:window.location.href='https://brave.com';";
document.body.appendChild(img);
```

```javascript
location.href="javascript:window.location.href='https://brave.com'";
```

Then I opened a console connected to my device remotely via Safari development tools and executed:
```javascript
$('body')[0].background="javascript:window.location.href='gmail.com'";
$('img')[0].dynsrc="javascript:console.log('hello');";
$('img')[0].lowsrc="javascript:console.log('hello');";
$('img')[0].dynsrc="javascript:console.log('hello');";
$('img')[0].src=" &#14;  javascript:console.log('hello');"; //gives a 404
```

etc..

For the text-field where they enter the URL for the bookmarklet, it won't pass validation, if it doesn't begin with exactly `"javascript:"`.
Only thing left to test would be the URL bar. I will test it and come back and edit this comment with the results :)

EDIT: Doesn't work in the URL bar either. It must begin with exactly `"javascript:"`

As for how Firefox does it: https://github.com/mozilla-mobile/firefox-ios/blob/master/Client/Frontend/Browser/BrowserViewController/BrowserViewController%2BWebViewDelegates.swift#L214

https://github.com/mozilla-mobile/firefox-ios/pull/4556

File: DataTests/BookmarkTests.swift [hunk 12]
```diff
@@ -583,4 +598,27 @@ class BookmarkTests: CoreDataTestCase {
         XCTAssertNotNil(record.children)
         XCTAssert(record.children!.isEmpty)
     }
+    
+    // Bookmarklet validation
+    private func validateBookmarklet(title: String?, url: String?) -> Bool {
```
Comment (iccub) [inline DataTests/BookmarkTests.swift]: This code is copied from `BookmarkDetailsView` if it changes there we might forget to update this test code.

Could we abstract it a bit? We could add an internal static method in `BookmarkDetailsView` and call it from both places(production and tests)

File: Shared/Extensions/URLExtensions.swift [hunk 13]
```diff
@@ -468,6 +468,39 @@ extension URL {
...
+    }
+    
+    public var bookmarkletCodeComponent: String? {
```
Comment (iccub) [inline Shared/Extensions/URLExtensions.swift:489]: Can you add a simple unit test here? One success assertion and one nil assertion

File: Shared/Extensions/URLExtensions.swift [hunk 14]
```diff
@@ -468,6 +468,39 @@ extension URL {
...
+        if self.isBookmarklet {
+            if let result = String(self.dropFirst("javascript:".count)).removingPercentEncoding {
+                return result.isEmpty ? nil : result
+            }
+        }
+        return nil
+    }
+    
+    public var bookmarkletURL: URL? {
```
Comment (iccub) [inline Shared/Extensions/URLExtensions.swift:498]: Can you add a simple unit test here? One success assertion and one nil assertion.

There's `NSURLExtensionsTests` class for that

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift [hunk 15]
```diff
@@ -69,7 +72,10 @@ class BookmarkDetailsView: AddEditHeaderView, BookmarkFormFieldsProtocol {
         [emptySpacer, faviconImageView, textFieldsStackView]
             .forEach(contentStackView.addArrangedSubview)
         
-        if let url = url, let favUrl = URL(string: url) {
+        var url = url
+        if url?.isBookmarklet ?? false {
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: I think in such unwrapped `if` conditions writing `url?.isBookmarklet == true` is easier to read

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift (hunk 6 above)
Comment (iccub) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: I tested keyboard types in other browsers and:
Safari and Firefox use url keyboard type while Chrome does not, I will talk to product team about it

**Review by iccub — CHANGES_REQUESTED** [review]

File: DataTests/BookmarkTests.swift (hunk 12 above)
Comment (Brandon-T) [reply DataTests/BookmarkTests.swift]: Fixed.. but the code was so tightly coupled to the view, I had to move a bunch of other code around as well otherwise the unit test gives a million errors trying to import.

Comment (iccub) [reply DataTests/BookmarkTests.swift]: Have you tried doing `@testable import Client` and just call the static method from there?
`@testable` allow to call internal methods from other frameworks

Comment (iccub) [reply DataTests/BookmarkTests.swift]: But I am not sure if calling that from Data framework tests works..

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift [hunk 16]
```diff
@@ -88,6 +93,19 @@ class BookmarkDetailsView: AddEditHeaderView, BookmarkFormFieldsProtocol {
     // MARK: - Delegate actions
     
     @objc func textFieldDidChange(_ textField: UITextField) {
-        delegate?.correctValues(validationPassed: validateFields())
+        if textField.text?.isBookmarklet ?? false {
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: Change to `== true` :)

**Review by iccub — CHANGES_REQUESTED** [review]
one minor change left
thanks for putting effort and making those unit tests

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift (hunk 16 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: Fixed.

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift (hunk 6 above)
Comment (iccub) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: So for now we should use the .url keyboard.
We will reconsider if there's people complaining about it

**Review by iccub — CHANGES_REQUESTED** [review]

**Review by iccub — APPROVED** [review]

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift (hunk 6 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarkDetailsView.swift]: @srirambv
